### PR TITLE
Checksum Cleanup

### PR DESCRIPTION
**Pull request by schisamo**
**NOTE**: This PR is dependent on a FK `ON DELETE` constraint change represented by the following pull-request in `chef-sql-schema`: https://github.com/opscode/chef-sql-schema/pull/5

Properly cleaning up redundant checksums requires 2 distintive steps of related work:
### 1. Delete Redundant Checksums in Database
- **chef_sql:delete_cookbook_version/1 now cleans up checksums** - Attempt to delete all checksum records associated with a particular cookbook version relying on the fact the RDBMS will throw a foreign key constraint error _if_ the checksum record is still needed. YesSQL!
- **chef_db/delete_cookbook_version/2 should expose same contract** - This function needs to return the same values, namely the 1 or 2 indicating whether the enclosing cookbook was deleted.  This flag is needed for proper authz cleanup.
### 2. Delete Checksum Data in S3

This logic has been implemented in `chef_objects` and is represented in this pull-request:
https://github.com/opscode/chef_objects/pull/6

### CODE REVIEW

File: priv/mysql_statements.config [hunk 1]
```diff
@@ -368,10 +368,16 @@
 {delete_sandbox_by_id,
  <<"DELETE FROM sandboxed_checksums WHERE sandbox_id = ?">>}.
 
+{find_checksum_by_id,
```
Comment (seth) [inline priv/mysql_statements.config]: Since we don't share checksums across orgs, I think this query needs to take an org_id as well. Same for the delete_checksums_by_id query, needs to limit to a specific org_id.

File: src/chef_db.erl [hunk 2]
```diff
@@ -745,7 +746,13 @@ update_cookbook_version(#context{}=Ctx, UpdatedCookbookVersion, ActorId) ->
                               CookbookVersion::#chef_cookbook_version{})
    -> {ok, 1 | 2} | not_found | {error, _}.
 delete_cookbook_version(#context{}=Ctx, #chef_cookbook_version{}=CookbookVersion) ->
-    delete_object(Ctx, delete_cookbook_version, CookbookVersion).
+    case delete_object(Ctx, delete_cookbook_version, CookbookVersion) of
```
Comment (seth) [inline src/chef_db.erl:750]: The returns value of `delete_object` here is feeling increasingly leaky. I wonder if there is some smaller piece of delete_object to factor out for reuse? More on this later. Also, as-is I'm confused because it looks like delete_object isn't going to handle the 3-element return properly.

File: src/chef_sql.erl [hunk 3]
```diff
@@ -833,12 +835,12 @@ delete_cookbook_version(#chef_cookbook_version{id=CookbookVersionId,
...
                                                name=Name}) ->
     case delete_checksums(OrgId, CookbookVersionId) of
-        {ok, 1 } ->
+        {ok, DeletedChecksums} ->
             case delete_object(delete_cookbook_version_by_id, CookbookVersionId) of
                 {ok, 1} ->
                     case delete_cookbook_if_last(OrgId, Name) of
                         {ok, N} ->
-                            {ok, N + 1};
+                            {ok, N+1, DeletedChecksums};
```
Comment (seth) [inline src/chef_sql.erl]: Let's try to think through what things would look like with a more structured return value. This might be a record or a proplist like `{ok, [{cookbook_delete, true|false}, {deleted_checksums, Checksums}]}`.

Also, it might be worth seeing if refactoring some of this with a helper function that handles some of the branching in funheads might improve the readability. I think it would be easier to read with a bit less nesting.

File: src/chef_sql.erl [hunk 4]
```diff
@@ -888,9 +890,21 @@ create_sandbox(#chef_sandbox{} = Sandbox) ->
...
 
-
 %% Checksum Operations
 
+%% @doc Helper function for testing checksum existence.
+-spec checksum_exists(Checksum :: binary()) ->
+                             boolean() | {error, term()}.
+checksum_exists(ChecksumId) ->
+    case sqerl:select(find_checksum_by_id, [ChecksumId], first_as_scalar, [checksum]) of
+        {ok, Checksum} when is_binary(Checksum) ->
```
Comment (seth) [inline src/chef_sql.erl]: minor style note: if you reorder the clauses you can remove the guard. Moreover, since you will only ever see Checksum as binary or none you could do:

```
{ok, Checksum} -> Checksum =/= none % or is_binary(Checksum)
```

to collapse the two clauses. It is fine as-is.

File: src/chef_sql.erl [hunk 5]
```diff
@@ -1325,16 +1339,41 @@ fetch_cookbook_authz(OrgId, CookbookName) ->
...
                        CookbookVersionId::binary()) -> {ok, 1 } | {error, _}.
 delete_checksums(OrgId, CookbookVersionId) ->
+    % retrieve a list of checksums before we delete the
+    % cookbook_version_checksums record
+    Checksums = fetch_cookbook_version_checksums(OrgId, CookbookVersionId),
     case sqerl:statement(delete_cookbook_checksums_by_orgid_cookbook_versionid,
                          [OrgId, CookbookVersionId]) of
         {ok, N} when is_integer(N) -> %% pretend there is 1
-            {ok, 1};
+            delete_checksums(Checksums);
```
Comment (seth) [inline src/chef_sql.erl]: Collapse these clauses since there is no difference in the follow on logic. So:

```
{ok, _} ->
    delete_checksums(Checksums)
```

File: src/chef_sql.erl [hunk 6]
```diff
@@ -1325,16 +1339,41 @@ fetch_cookbook_authz(OrgId, CookbookName) ->
...
+%% foreign_key constraint errors assuming the checksum is still associated with
+%% another cookbook_version_checksum record.  Returns a list of deleted checksum
+%% ids for further upstream processing(ie delete the checksums from S3).
+delete_checksums(Checksums) ->
+    DeletedChecksums = lists:foldl(fun(Checksum, Acc) ->
+                            case sqerl:statement(delete_checksum_by_id, [Checksum]) of
+                                {ok, N} when is_integer(N) -> %% pretend there is 1
+                                    [Checksum|Acc];
+                                {foreign_key, _} ->
+                                    %% The checksum may still be associated with
```
Comment (seth) [inline src/chef_sql.erl]: s/may/are/

File: src/chef_sql.erl [hunk 7]
```diff
@@ -1325,16 +1339,41 @@ fetch_cookbook_authz(OrgId, CookbookName) ->
...
+                                    %% another cookbook version record which is OK!
+                                    Acc;
+                                {error, _Error} ->
```
Comment (seth) [inline src/chef_sql.erl]: We should think about and add comments here on what we want to do if an error occurs. Why do you think swallowing the error is the right thing here? We should at least add some logging.

File: src/chef_sql.erl [hunk 8]
```diff
@@ -1325,16 +1339,41 @@ fetch_cookbook_authz(OrgId, CookbookName) ->
...
+            delete_checksums(Checksums);
         {error, Error} ->
             {error, Error}
     end.
 
+%% @doc Deletes a list of checksums from the checksums table.  Happily swallows
+%% foreign_key constraint errors assuming the checksum is still associated with
+%% another cookbook_version_checksum record.  Returns a list of deleted checksum
+%% ids for further upstream processing(ie delete the checksums from S3).
+delete_checksums(Checksums) ->
```
Comment (seth) [inline src/chef_sql.erl]: Pretty sure this will need an OrgId arg.

File: src/chef_sql.erl [hunk 9]
```diff
@@ -888,9 +890,21 @@ create_sandbox(#chef_sandbox{} = Sandbox) ->
 delete_sandbox(SandboxId) when is_binary(SandboxId) ->
     delete_object(delete_sandbox_by_id, SandboxId).
 
-
 %% Checksum Operations
 
+%% @doc Helper function for testing checksum existence.
+-spec checksum_exists(Checksum :: binary()) ->
+                             boolean() | {error, term()}.
+checksum_exists(ChecksumId) ->
```
Comment (christophermaier) [inline src/chef_sql.erl]: It looks like this function is just required for testing, and isn't used in the actual operating code, right?

In that case, it would be a good idea to pull it into the itests file or (better) pull out a separate testing helper module and stick it there (since the itest file is getting ridiculously large).  You can add the prepared statement into Sqerl when the tests are set up (see what I did with the cookbook test setup)

File: priv/mysql_statements.config (hunk 1 above)
Comment (christophermaier) [inline priv/mysql_statements.config]: Yup

File: src/chef_sql.erl (hunk 7 above)
Comment (christophermaier) [inline src/chef_sql.erl]: Agree that comments and logging would be good.  I'm not sure what we could meaningfully do in an error situation, though, given that we've already deleted the cookbook version at this point, as well as (potentially) some number of checksums.  The horses have kind of left the barn at that point.

Now, if sqerl had transactions....

File: src/chef_sql.erl [hunk 10]
```diff
@@ -1325,16 +1339,41 @@ fetch_cookbook_authz(OrgId, CookbookName) ->
...
+                                    %% The checksum may still be associated with
+                                    %% another cookbook version record which is OK!
+                                    Acc;
+                                {error, _Error} ->
+                                    Acc
+                            end
+                      end,
+                      [],
+                      Checksums),
+    {ok, lists:reverse(DeletedChecksums)}.
```
Comment (christophermaier) [inline src/chef_sql.erl]: I don't think list reversal is needed here, given that checksums don't have any inherent order.

File: src/chef_db.erl (hunk 2 above)
Comment (christophermaier) [inline src/chef_db.erl:750]: The intention of bubbling up the deleted checksums was to separate S3/Bookshelf related operations (nuking the actual files) from the database operations (nuking our /knowledge/ of the files).  @schisamo and I talked about it, and this seemed like the most logical place to coordinate those activities.

File: src/chef_sql.erl (hunk 3 above)
Comment (christophermaier) [inline src/chef_sql.erl]: Maybe a small record... not sure I like a proplist for this, since you can't really pattern match on it.  I'm OK with the return value as it is, though (but I'd definitely want to refactor if more data were added later).

A de-nesting might be nice, but the nesting wasn't introduced here, so I'd be OK with postponing that particular refactor until later.

File: src/chef_db.erl (hunk 2 above)
Comment (christophermaier) [inline src/chef_db.erl:750]: And yeah, delete_object looks like it needs a little 3-argument love.

**Comment by schisamo** [comment]
^^ @christophermaier @seth 

I still need to sort out the return record from `chef_sql:delete_cookbook_version/1` but I made some progress based on your feedback.

**Comment by christophermaier** [comment]
So far, so good!

**Comment by schisamo** [comment]
^^ @christophermaier @seth got the new record worked in.  Ratholed slightly fixing numerous dialyzer errors.  Will be starting on the actual s3 deletion logic in the AM.

**Comment by schisamo** [comment]
^^ @seth @christophermaier final feedback when you have a sec.

File: itest/itest.erl [hunk 11]
```diff
@@ -1598,32 +1602,69 @@ delete_cookbook_version_checksums() ->
                                            {CookbookVersion#chef_cookbook_version.major,
                                             CookbookVersion#chef_cookbook_version.minor,
                                             CookbookVersion#chef_cookbook_version.patch}}),
-    ?assertEqual({ok, 2}, %% Last version of this cookbook, so deleting the cookbook as well
-                 chef_sql:delete_cookbook_version(Got)),
-    %% TODO - how to check if the checksums have been deleted ?
+
+    % Verify all checksums exist
```
Comment (seth) [inline itest/itest.erl]: Minor, but prefer `%%` for line comments

File: src/chef_sql.erl [hunk 12]
```diff
@@ -1322,18 +1326,45 @@ fetch_cookbook_authz(OrgId, CookbookName) ->
...
+    lists:foldl(fun(Checksum, Acc) ->
+            case sqerl:statement(delete_checksum_by_id, [OrgId, Checksum]) of
+                {ok, N} when is_integer(N) -> %% pretend there is 1
+                    [Checksum|Acc];
+                {foreign_key, _} ->
+                    %% The checksum may still be associated with
+                    %% another cookbook version record which is OK!
+                    Acc;
+                {error, Error} ->
+                    error_logger:error_msg("The following error occured
```
Comment (seth) [inline src/chef_sql.erl]: Since this is an internal error message, I'd shorten this to indicate the Mod and Fun and details specific to the error.

File: src/chef_sql.erl [hunk 13]
```diff
@@ -832,13 +834,16 @@ update_cookbook_version_checksums(#chef_cookbook_version{ id        = Id,
...
             case delete_object(delete_cookbook_version_by_id, CookbookVersionId) of
                 {ok, 1} ->
                     case delete_cookbook_if_last(OrgId, Name) of
-                        {ok, N} ->
-                            {ok, N + 1};
+                        {ok, 0} ->
+                            DeleteResponse#chef_db_cb_version_delete{cookbook_delete = false};
+                        {ok, 1} ->
+                            DeleteResponse#chef_db_cb_version_delete{cookbook_delete = true};
                         Error = {error, _} ->
```
Comment (seth) [inline src/chef_sql.erl]: This is ok, but I think might be easier to read with a helper function and not as a many "exits". For example:

```
delete_cookbook_version(#chef_cookbook_version{id=CookbookVersionId,
                                               org_id=OrgId,
                                               name=Name}) ->
    case delete_cookbook_version_checksums(OrgId, CookbookVersionId) of
        {ok, DeletedChecksums} ->
            CookbookDelete = case delete_object(delete_cookbook_version_by_id, CookbookVersionId) of
                                 {ok, 1} ->
                                     case delete_cookbook_if_last(OrgId, Name) of
                                         {ok, 0} -> false;
                                         {ok, 1} -> true;
                                         Error = {error, _} -> Error
                                     end;
                                 {error, Reason} ->
                                     {error, Reason}
                             end,
            dcv_result_or_error(CookbookDelete, DeletedChecksums);
        {error, Reason} ->
            {error, Reason}
    end.

dcv_result_or_error({error, Reason}, _) ->
    {error, Reason};
dcv_result_or_error(CookbookDelete, DeletedChecksums) ->
    #chef_db_cb_version_delete{deleted_checksums = DeletedChecksums,
                               cookbook_delete = CookbookDelete}.
```

**Comment by schisamo** [comment]
This has already been merged.
